<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN"
    "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<meta name="generator" content="AsciiDoc 8.2.6" />
<style type="text/css">
/* Debug borders */
p, li, dt, dd, div, pre, h1, h2, h3, h4, h5, h6 {
/*
  border: 1px solid red;
*/
}

body {
  margin: 1em 5% 1em 5%;
}

a {
  color: blue;
  text-decoration: underline;
}
a:visited {
  color: fuchsia;
}

em {
  font-style: italic;
  color: navy;
}

strong {
  font-weight: bold;
  color: #083194;
}

tt {
  color: navy;
}

h1, h2, h3, h4, h5, h6 {
  color: #527bbd;
  font-family: sans-serif;
  margin-top: 1.2em;
  margin-bottom: 0.5em;
  line-height: 1.3;
}

h1, h2, h3 {
  border-bottom: 2px solid silver;
}
h2 {
  padding-top: 0.5em;
}
h3 {
  float: left;
}
h3 + * {
  clear: left;
}

div.sectionbody {
  font-family: serif;
  margin-left: 0;
}

hr {
  border: 1px solid silver;
}

p {
  margin-top: 0.5em;
  margin-bottom: 0.5em;
}

ul, ol, li > p {
  margin-top: 0;
}

pre {
  padding: 0;
  margin: 0;
}

span#author {
  color: #527bbd;
  font-family: sans-serif;
  font-weight: bold;
  font-size: 1.1em;
}
span#email {
}
span#revision {
  font-family: sans-serif;
}

div#footer {
  font-family: sans-serif;
  font-size: small;
  border-top: 2px solid silver;
  padding-top: 0.5em;
  margin-top: 4.0em;
}
div#footer-text {
  float: left;
  padding-bottom: 0.5em;
}
div#footer-badges {
  float: right;
  padding-bottom: 0.5em;
}

div#preamble,
div.tableblock, div.imageblock, div.exampleblock, div.verseblock,
div.quoteblock, div.literalblock, div.listingblock, div.sidebarblock,
div.admonitionblock {
  margin-right: 10%;
  margin-top: 1.5em;
  margin-bottom: 1.5em;
}
div.admonitionblock {
  margin-top: 2.5em;
  margin-bottom: 2.5em;
}

div.content { /* Block element content. */
  padding: 0;
}

/* Block element titles. */
div.title, caption.title {
  color: #527bbd;
  font-family: sans-serif;
  font-weight: bold;
  text-align: left;
  margin-top: 1.0em;
  margin-bottom: 0.5em;
}
div.title + * {
  margin-top: 0;
}

td div.title:first-child {
  margin-top: 0.0em;
}
div.content div.title:first-child {
  margin-top: 0.0em;
}
div.content + div.title {
  margin-top: 0.0em;
}

div.sidebarblock > div.content {
  background: #ffffee;
  border: 1px solid silver;
  padding: 0.5em;
}

div.listingblock {
  margin-right: 0%;
}
div.listingblock > div.content {
  border: 1px solid silver;
  background: #f4f4f4;
  padding: 0.5em;
}

div.quoteblock > div.content {
  padding-left: 2.0em;
}

div.attribution {
  text-align: right;
}
div.verseblock + div.attribution {
  text-align: left;
}

div.admonitionblock .icon {
  vertical-align: top;
  font-size: 1.1em;
  font-weight: bold;
  text-decoration: underline;
  color: #527bbd;
  padding-right: 0.5em;
}
div.admonitionblock td.content {
  padding-left: 0.5em;
  border-left: 2px solid silver;
}

div.exampleblock > div.content {
  border-left: 2px solid silver;
  padding: 0.5em;
}

div.verseblock div.content {
  white-space: pre;
}

div.imageblock div.content { padding-left: 0; }
div.imageblock img { border: 1px solid silver; }
span.image img { border-style: none; }

dl {
  margin-top: 0.8em;
  margin-bottom: 0.8em;
}
dt {
  margin-top: 0.5em;
  margin-bottom: 0;
  font-style: normal;
}
dd > *:first-child {
  margin-top: 0.1em;
}

ul, ol {
    list-style-position: outside;
}
div.olist > ol {
  list-style-type: decimal;
}
div.olist2 > ol {
  list-style-type: lower-alpha;
}

div.tableblock > table {
  border: 3px solid #527bbd;
}
thead {
  font-family: sans-serif;
  font-weight: bold;
}
tfoot {
  font-weight: bold;
}

div.hlist {
  margin-top: 0.8em;
  margin-bottom: 0.8em;
}
div.hlist td {
  padding-bottom: 15px;
}
td.hlist1 {
  vertical-align: top;
  font-style: normal;
  padding-right: 0.8em;
}
td.hlist2 {
  vertical-align: top;
}

@media print {
  div#footer-badges { display: none; }
}

div#toctitle {
  color: #527bbd;
  font-family: sans-serif;
  font-size: 1.1em;
  font-weight: bold;
  margin-top: 1.0em;
  margin-bottom: 0.1em;
}

div.toclevel1, div.toclevel2, div.toclevel3, div.toclevel4 {
  margin-top: 0;
  margin-bottom: 0;
}
div.toclevel2 {
  margin-left: 2em;
  font-size: 0.9em;
}
div.toclevel3 {
  margin-left: 4em;
  font-size: 0.9em;
}
div.toclevel4 {
  margin-left: 6em;
  font-size: 0.9em;
}
/* Workarounds for IE6's broken and incomplete CSS2. */

div.sidebar-content {
  background: #ffffee;
  border: 1px solid silver;
  padding: 0.5em;
}
div.sidebar-title, div.image-title {
  color: #527bbd;
  font-family: sans-serif;
  font-weight: bold;
  margin-top: 0.0em;
  margin-bottom: 0.5em;
}

div.listingblock div.content {
  border: 1px solid silver;
  background: #f4f4f4;
  padding: 0.5em;
}

div.quoteblock-content {
  padding-left: 2.0em;
}

div.exampleblock-content {
  border-left: 2px solid silver;
  padding-left: 0.5em;
}

/* IE6 sets dynamically generated links as visited. */
div#toc a:visited { color: blue; }

/* Because IE6 child selector is broken. */
div.olist2 ol {
  list-style-type: lower-alpha;
}
div.olist2 div.olist ol {
  list-style-type: decimal;
}
</style>
<title>Extract Likely Coding Sequences from Trinity Transcripts</title>
</head>
<body>
<div id="header">
<h1>Extract Likely Coding Sequences from Trinity Transcripts</h1>
</div>
<div id="preamble">
<div class="sectionbody">
<div class="para"><p>Likely coding regions can be extracted from Trinity transcripts using a set of utilities included in the Trinity software distribution.  The system works as follows:</p></div>
<div class="ilist"><ul>
<li>
<p>
the longest ORF is identified within the Trinity transcript (if strand-specific, this is restricted to the top strand).
</p>
</li>
<li>
<p>
of all the longest ORFs extracted, a subset corresponding to the very longest ones (and most likely to be genuine) are identified and used to parameterize a Markov model based on hexamers.  These likely coding sequences are randomized to provide a sequence composition corresponding to non-coding sequence.
</p>
</li>
<li>
<p>
all longest ORFs found are scored according to the Markov Model (log likelihood ratio based on coding/noncoding) in each of the six possible reading frames. If the putative ORF proper coding frame scores positive and is highest of the other presumed wrong reading frames, then that ORF is reported.
</p>
</li>
</ul></div>
<div class="para"><p>The scoring of ORFs is largely based on that described for <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC310871/?tool=pubmed">GeneID</a>, and the ORF selection process was added here.</p></div>
</div>
</div>
<h2 id="_extracting_best_orfs">Extracting Best ORFs</h2>
<div class="sectionbody">
<div class="para"><p>Extracting likely coding regions from Trinity transcripts can be done using (beware, set up for strand-specific data by default, use -B for both strands):</p></div>
<div class="literalblock">
<div class="content">
<pre><tt>TRINITY_RNASEQ_ROOT/Analysis/Coding/transcripts_to_best_scoring_ORFs.pl</tt></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><tt>with options:</tt></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><tt>############################# Options ###################################################
#
# Required:
#
# -t transcripts.fasta
#
#  Optional:
#
# -m minimum protein length (default: 100)
#
# -G genetic code (default: universal, options: Euplotes, Tetrahymena, Candida, Acetabularia)
#
# -h print this option menu and quit
# -v verbose
#
# -C complete ORFs only ********
# -B examine both strands (default: only top strand)
# -T top longest ORFs to train Markov Model (hexamer stats) (default: 500)
#
##############################################################################################</tt></pre>
</div></div>
</div>
<h2 id="_sample_data">Sample Data</h2>
<div class="sectionbody">
<div class="para"><p>Sample data and example execution are provided at</p></div>
<div class="literalblock">
<div class="content">
<pre><tt>TRINITY_RNASEQ_ROOT/sample_data/test_Trinity_Coding_Extraction</tt></pre>
</div></div>
<div class="para"><p>Just run <em>runMe.sh</em> to run the process.</p></div>
<div class="para"><p>Output files include:</p></div>
<div class="literalblock">
<div class="content">
<pre><tt>-rw-r--r--  1 bhaas  admin  106682 Oct 29 12:54 best_candidates.pep
-rw-r--r--  1 bhaas  admin  313901 Oct 29 12:54 best_candidates.cds
-rw-r--r--  1 bhaas  admin    5659 Oct 29 12:54 best_candidates.bed</tt></pre>
</div></div>
<div class="para"><p>The .bed file can be loaded into IGV for viewing along with the additional Trinity assembly and alignment data.</p></div>
</div>
<div id="footer">
<div id="footer-text">
Last updated 2011-10-29 12:55:58 EDT
</div>
</div>
</body>
</html>
